### ISSUE TITLE
Kubernetes pods running as root and privileged

### ISSUE THREAD

**Issue by MarkDeckert** [gravitee-io/issues#4078]
Kubernetes deployments should be brought up to at least bare minimum container security best practices. By default everything except mongo runs as root and in the case of the ElasticSearch deployed by helm explicitly runs in privileged mode.

This is highly problematic in practice since this gives the application the ability to modify or destroy critical components on the host -- especially for services which by design will be open to external traffic. Also this will not pass review in enterprises that have basic security requirements that applications should not be running with unnecessarily escalated privileges.

None of the components require root or privileged. It would be highly uncommon for any application to need this level of access. Certainly it makes it easy to build when everything has complete unfettered access to the underlying system but no production deployment should allow this.

### Expected Behavior
All components running with least privileges necessary to function.

### Current Behavior
Docker images apparently run as root.
Kubernetes yaml is missing securityContext values.

### Possible Solution
1. Docker images should be built using a non-0 USER
2. Generated Kubernetes yaml should include securityContext sections specifying at the minimum for each container:
```
securityContext:
  privileged: false
  runAsUser: (something-not-0)
```
3. Containers listening on ports lower than 1024 in helm charts should default to a higher port (ie 8080 instead of 80)

**Comment by brasseld** [comment]
Hi @MarkDeckert 

Thanks for pointing this.

I completely agree with your vision. For the moment, you can define an helm value to setup the securityContext: https://github.com/gravitee-io/helm-charts/blob/master/apim/3.x/templates/ui/ui-deployment.yaml#L48

This shall be subject to change for UI components which are running Nginx on port 80 / 443. As you noticed, we must change default port to 8080 / 8043 to setup securityContext by default.

**Comment by MarkDeckert** [comment]
After further testing it appears there may also be some issues with the helm chart and these options in particular.  Let me know if I should file a separate issue though.

### SecurityContext section
When I set uri.runAsNonRoot to true and do a helm template - I get the following error:
`Error: YAML parse error on apim3/templates/ui/ui-deployment.yaml: error converting YAML to JSON: yaml: line 36: mapping values are not allowed in this context`
If I set it to false there is no issue - although obviously the incorrect settings are not applied.

For now I can just set the runAsUser to 1001 manually in the resulting yaml.

### InternalPort values

When i set ui.service.internalPort: 8080, I get a different issue. Although in the resulting yaml the container port of the management UI does get set to 8080, the pod still fail to launch.  After examining the Dockerfile/configs of the image it appears in order for this to work correctly the helm chart needs to also create an environment variable for HTTP_PORT and set it to 8080 since this is how confd is going to set the nginx config.

Although that's actually not good enough since the nginx config also needs an https port.  An HTTPS_PORT environment variable must also be set to something higher, but obviously not 8080, so really the helm chart can't reuse the same value and instead needs an additional value option for the HTTPS port ("internalHTTPSPort?). I prefer 8443 so that's what I've set it to manually in the resulting yaml as a workaround and now nginx can start.

I should note here that portal.service.internalPort would need to work the same way, although the documentation is also missing info about portal settings.  I only discovered these by reading through the helm charts.

**Comment by brasseld** [comment]
Hi @MarkDeckert 

We will change default HTTP / HTTPS port for all UI images: https://github.com/gravitee-io/issues/issues/4094

And securityContext will be enabled by default once release 3.1 is done.

**Comment by aelamrani** [comment]
Images has been updated, waiting for your feedback.

**Comment by MarkDeckert** [comment]
I tested "3.0.5" and "latest" images without the additional HTTP_PORT and HTTPS_PORT environment variables and indeed that is fixed.  Thank you!  Looking forward to release 3.1 with the securityContext. That will fix issues with PodSecurityPolicy.

Looks like current helm chart is still on 3.0.4 release so I'll leave this open for now in case others come across the issue in current release.  Feel free to close this issue now if you prefer.

**Comment by brasseld** [comment]
Hi @MarkDeckert 

I've created a PR for this. Do you think you can have a look into it?

https://github.com/gravitee-io/helm-charts/pull/63/files

Regards,

**Comment by ingvarbs** [comment]
Hi @brasseld 

Is it possible to make  runAsUser configurable, so that one can either specify the UID and / or specify wether  runAsUser is specified at all?

This will make it run on openshift/okd ( openshift spesifies a range of UIDs for each namespace, and unprivileged pods need to run as a user with UID from this range. If runAsUser isn't spesified openshift selects one from this range)

Regards,
Ingvar

**Comment by brasseld** [comment]
Hi @ingvarbs,

Only the uid or also the gid ?

**Comment by ingvarbs** [comment]
Hi @brasseld .
this applies to gid as well (GID is treated the same way in openshft, with a range per namespace, per default with same numbers as uid),

**Comment by brasseld** [comment]
Makes sense,

I'll manage this from an other Pull Request as there are already some works about OpenShift:
https://github.com/gravitee-io/helm-charts/pull/62

**Comment by MarkDeckert** [comment]
@brasseld Sorry about my slow replies.  I left a comment on https://github.com/gravitee-io/helm-charts/pull/63 but I think this will fully resolve the issue.

**Comment by MarkDeckert** [comment]
Sorry about the accidental closed/reopen.  Please close once merged.
